lib/std/io/Reader.zig

```diff
@@ -324,20 +324,47 @@ pub fn isBytes(self: Self, slice: []const u8) anyerror!bool {
     return matches;
 }
 
+/// Read a struct from the stream.
+/// Only packed and extern structs are supported.
+/// Packed structs must have a `@bitSizeOf` that is a multiple of eight.
 pub fn readStruct(self: Self, comptime T: type) anyerror!T {
-    // Only extern and packed structs have defined in-memory layout.
-    comptime assert(@typeInfo(T).@"struct".layout != .auto);
-    var res: [1]T = undefined;
-    try self.readNoEof(mem.sliceAsBytes(res[0..]));
-    return res[0];
+    switch (@typeInfo(T).@"struct".layout) {
```

Comment:
this code

```zig
test "readStruct compile error for u8" {
    const buf = [3]u8{ 11, 12, 13 };
    var fis = std.io.fixedBufferStream(&buf);
    const reader = fis.reader();
    _ = try reader.readStruct(u8);
}

```

produces the following compile error:

```
zig test lib/std/std.zig --zig-lib-dir lib --test-filter "io."
lib/std/io/Reader.zig:331:25: error: access of union field 'struct' while field 'int' is active
    switch (@typeInfo(T).@"struct".layout) {
            ~~~~~~~~~~~~^~~~~~~~~~
lib/std/builtin.zig:268:18: note: union declared here
pub const Type = union(enum) {
```

Which is a bit annoying, since it doesn't tell me where I went wrong

Reply:
Yeah, not a great error message.  I'm not sure if this is something that should be fixed here with an explicit check, or left for Zig to fix more generically somehow.  Seems fine to me to leave this as-is for now, if no one more knowledgeable chimes in.

Reply:
I personally also prefer `@compileError` stating the intent, however there's been precedent of rejecting such lines from `std` for keeping the code shorter/focused: https://github.com/ziglang/zig/pull/20748